### ISSUE TITLE
Docs: Official Documentation For Custom Fonts

### ISSUE THREAD

**Issue by kylemh** [vercel/next.js#35582]
### What is the improvement or update you wish to see?

Official documentation on how to deal with custom fonts should be listed within the Font Optimization page at the very top and it should very clearly explain how I can best use custom fonts on my Next.js application.

### Is there any context that might help us understand?

Here's the journey for me, today:

I google: "Next.js custom font" and am given this as the top result:
https://nextjs.org/docs/basic-features/font-optimization

I happen to know that this only works for Google fonts, but somebody else might not. They'll be very confused until they scroll _very_ far down the page to see a tiny aside listed at the bottom of the page that says:

> Note: Font Optimization does not currently support self-hosted fonts.

So, self-hosted font loading strategies are not described anywhere in the Next.js documentation leaving self-hosted font users floundering around the web.

Now what?

- Searching issues yields a lot of irrelevant or unanswered tickets.
- [In discussions (with the search: "self-hosted font")](https://github.com/vercel/next.js/discussions?discussions_q=self-hosted+font)
- [In discussions (with the search: "custom font")](https://github.com/vercel/next.js/discussions?discussions_q=custom+font)
- [This popular stack overflow post](https://stackoverflow.com/questions/61909298/how-to-use-self-hosted-fonts-face-using-nextjs)

Somewhere along this wild ride, I'm directed to this fabulous website from @leerob: [https://leerob.io/blog/fonts](https://leerob.io/blog/fonts)

I'm on the hunt to remove layout shift, so - after reading this article - I am preloading all my font assets from the public folder in `_document` AND I'm using `font-display: optional` as outlined. All seems sorted, but once I deploy my build, I see that - very often - the custom font is never used. It just stays as Arial.

In devtools, I see the following error:

```
The resource <URL> was preloaded using link preload but not used within a few seconds from the window's load event. Please make sure it has an appropriate `as` value and it is preloaded intentionally.
website.com/:1 The resource https://website.com/public/custom-font.woff2 was preloaded using link preload but not used within a few seconds from the window's load event. Please make sure it has an appropriate `as` value and it is preloaded intentionally.
```

This leads me to [a Next.js issue which dismisses the warning](https://github.com/vercel/next.js/issues/6517#issuecomment-469063892) and 2 popular Stack overflow issues that don't have a solution [[1](https://stackoverflow.com/questions/49674092/preloading-font-with-rel-preload) and [2](https://stackoverflow.com/questions/55819473/chrome-unused-preload-warning-for-an-icon-font-that-is-used)]

So... I'm feeling stuck with either buying into layout shift or buying into my users seeing Arial like 30% of the time and almost 100% of the time on a reload in-app.

**Comment by kylemh** [comment]
Hey @molebox (not sure who I should tag, sorry if I shouldn't have tagged you).

So, IMO Lee's docs get you 80% of the way. The rest that needs to be documented is creating a web worker to cache the fonts on Chrome since it doesn't automatically cache fonts. Mostly covered [here](https://gomakethings.com/improving-web-font-performance-with-service-workers/).

Here's the web worker I ended up with for [capsule.com](https://capsule.com):
```js
// @ts-nocheck
/** Must be ES5 */

const fontFiles = [
  '/public/font-file.woff2',
  // etc.
];

// On service worker install, cache files
self.addEventListener('install', function (event) {
  event.waitUntil(
    caches.open('core').then(function (cache) {
      fontFiles.forEach(function (file) {
        cache.add(new Request(file));
        return;
      });

      return;
    })
  );
});

// Listen for requests
self.addEventListener('fetch', function (event) {
  const request = event.request;

  /** @see https://stackoverflow.com/a/49719964 */
  if (event.request.cache === 'only-if-cached' && event.request.mode !== 'same-origin') return;

  // Fonts: Offline First
  if (request.url.includes('public/font-name-prefix')) {
    event.respondWith(
      caches.match(request).then(function (response) {
        return (
          response ||
          fetch(request).then(function (response) {
            // Return the requested file
            return response;
          })
        );
      })
    );
  }
});

```

**Comment by GreetVdL** [comment]
Hi @kylemh I had the same problem trying to preload my self-hosted fonts, but I think I've found a sloution: https://stackoverflow.com/questions/71690880/how-preload-self-hosted-fonts-with-next-js-console-warning-the-resource-was-no/71732470#71732470

**Comment by kylemh** [comment]
I did not have the problem you linked. There were no hashed file names to account for.

**Comment by GreetVdL** [comment]
I wasn't sure what the problem was, but it seems to be fixed this way. The fonts do preload and the console warnings are gone.

**Comment by kylemh** [comment]
In your Stack Overflow issue, you say: "the problem seems to be".  Did you confirm that the file names showed up as hashed in your public folder and/or that the file URL value changed in your CSS?

I ask because I think your fix may have worked for another reason. Google doesn't automatically cache font type resources, but perhaps it does automatically cache "asset/resource" types. Then, your webpack fix would be an alternative to the service worker.

**Comment by GreetVdL** [comment]
I tried to figure it out, but I'm a bit confused about what fixed the issue for me now. I must say I'm still quite new to web development. Apparently the problem was that in my css file I had used a relative path to my public/fonts folder. This created a hashed font file in the .next folder which produced the console warnings. I then changed my paths to be absolute but the hashed font file remained in the .next folder. Meanwhile after the webpack addition the warnings disappeared, so I thought that was the solution, but apparently the preloading now just works (without the webpack addition in next.config) if I use absolute paths in the_app.js and css file. The console warnings are gone. So I think you only get the console warnings when you use relative paths and you don't need the webpack part to solve the issue.

**Comment by kylemh** [comment]
So, if you do that, just note that Chrome doesn't automatically cache font files. So, if it's swapping in time, you're simply getting lucky.

**Comment by vasco3** [comment]
> Using custom, self-hosted fonts is now well documented. https://nextjs.org/docs/pages/building-your-application/optimizing/fonts

can this be used without SWC?

**Comment by leerob** [comment]
I believe yes, are you seeing issues using Babel?

**Comment by vasco3** [comment]
@leerob yes I am seeing this error https://nextjs.org/docs/messages/babel-font-loader-conflict
I would like to use some babel plugins for pipe operators and do expressions. Is there any other optimal recommended way to load fonts in Next?

**Comment by leerob** [comment]
Ah, my apologies. I did not realize that. Which babel plugins are you using?

**Comment by vasco3** [comment]
@leerob I was trying these
- @babel/plugin-proposal-do-expressions
- @babel/plugin-proposal-partial-application
- @babel/plugin-syntax-pipeline-operator

but I couldn't manage to get the pipe operator to work. So I rolledback to SWC
